### ISSUE TITLE
Lombok: JavaParser fails with  StringIndexOutOfBoundsException when lombok processor is enabled

### ISSUE THREAD

**Issue by amishra-u** [openrewrite/rewrite#4781]
## What version of OpenRewrite are you using?
- OpenRewrite 8.41.2

## What is the smallest, simplest way to reproduce the problem?
I haven't debugged the exact root cause yet. the issue seems to occur when annotations are added or removed in the source code. This appears to result in incorrect character positions for the cursor. 
(Will add more detail about the issue after debugging)

```java
  @Getter
  @Builder
  public static class Request {
    @Nonnull private final List<TransportJobId> transportJobIds;
    @Nullable private final UnixTimeMillis estimatedReadyAt;
    @Nonnull private final RegionId regionId;
    @Nonnull private final boolean wasTriggeredFromFulfillmentOrder;
  }
```

## What did you expect to see?
A valid compilation unit should be generated for the source file.

## What did you see instead?
A [ParseError](https://github.com/openrewrite/rewrite/blob/a98d83d698dda9971c93a0767b7f85d5866f306d/rewrite-core/src/main/java/org/openrewrite/tree/ParseError.java) instance is generated for the source file instead.

## What is the full stack trace of any errors you encountered?

```
java.lang.StringIndexOutOfBoundsException: begin 5327, end 5315, length 5434
  java.base/java.lang.String.checkBoundsBeginEnd(String.java:3319)
  java.base/java.lang.String.substring(String.java:1874)
  org.openrewrite.java.isolated.ReloadableJava17ParserVisitor.visitAnnotation(ReloadableJava17ParserVisitor.java:143)
  org.openrewrite.java.isolated.ReloadableJava17ParserVisitor.visitAnnotation(ReloadableJava17ParserVisitor.java:72)
  com.sun.tools.javac.tree.JCTree$JCAnnotation.accept(JCTree.java:2601)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
  org.openrewrite.java.isolated.ReloadableJava17ParserVisitor.convert(ReloadableJava17ParserVisitor.java:1604)
  ...

```

## Are you interested in [contributing a fix to OpenRewrite]
Yes.

**Comment by knutwannheden** [comment]
This is basically the reason we haven't enabled the Lombok processor yet. It appears like it doesn't always adjust the offsets in the AST correctly. I suspect that us a bug in Lombok. So either we need to get it fixed there or we need to adjust our parser so that it tracks the cursor in the file without relying on the AST.

**Comment by knutwannheden** [comment]
I think we are actually very close to have this working, we just didn't find the time to complete it.

**Comment by knutwannheden** [comment]
If you feel like taking a look at it, I believe that this is where we need to fix things to no longer rely on the position from the AST: https://github.com/openrewrite/rewrite/blob/abb80e16f579275811d4166bfd6387f504b25b16/rewrite-java-17/src/main/java/org/openrewrite/java/isolated/ReloadableJava17ParserVisitor.java#L1682

**Comment by timtebeek** [comment]
Thanks for the runnable example @amishra-u ! That should help zero in on the issue. Did you by chance look at a fix as well?

**Comment by amishra-u** [comment]
> Thanks for the runnable example @amishra-u ! That should help zero in on the issue. Did you by chance look at a fix as well?

Yes, I’ve identified the root cause. It’s not an OpenRewrite bug, but rather an issue with the Lombok annotation processor. I added a workaround in OpenRewrite, which resolved the problem. I’m also looking into fixing the underlying issue in Lombok itself and will publish a PR soon.

Lombok incorrectly updates the annotation type’s position on a field to the variable’s position in certain scenarios. If you closely look after image it has correct position for annotation but only the annotation type position is updated to the variable’s position.
Before annotation processor:
![Screenshot 2024-12-18 at 3 23 35 AM](https://github.com/user-attachments/assets/ee75e455-4ee7-4383-9c80-f7dc1db95ef5)
After annotation processor:
![Screenshot 2024-12-18 at 3 24 36 AM](https://github.com/user-attachments/assets/2684ad11-5131-476b-8305-851dfdec595a)

**Comment by timtebeek** [comment]
Awesome work! Thanks for diving in and posting the outcome here. Keep us posted on any progress with regards to a workaround or fix upstream.

**Comment by amishra-u** [comment]
This method is buggy, it doesn't make deep copy. 
https://github.com/projectlombok/lombok/blob/master/src/core/lombok/javac/handlers/JavacHandlerUtil.java#L2111

**Comment by knutwannheden** [comment]
This is also what I observed, but I never analyzed it in detail. Fixing this upstream would be awesome! Hopefully that will then address all such errors in Lombok.

**Comment by amishra-u** [comment]
Publish the draft PR with the unit test.
@timtebeek, could you please help verify it with OpenRewrite, even before this change is released?

**Comment by timtebeek** [comment]
hi @amishra-u ; sure! I don't quite know how to install a custom build of Lombok to your local Maven repository, but perhaps you already have that part figured out from your work there.

The version of Lombok we use is set (and can be overwritten) in the rewrite-build-gradle-plugin:
https://github.com/openrewrite/rewrite-build-gradle-plugin/blob/a6a7fd4ef41b0e123df09bc036a20c697b8cc485/src/main/java/org/openrewrite/gradle/RewriteJavaPlugin.java#L72-L77
If you then publish the rewrite-build-gradle-plugin to Maven local you can update the versions used here to use `latest.integration`:
https://github.com/openrewrite/rewrite/blob/a8946ac9c7b6cbe0b6ac1c540d5e5cfc4e2d0919/build.gradle.kts#L1-L5

I hope that helps you already there. Feel free to share the steps you did for the local installation of Lombok here as well.

**Comment by amishra-u** [comment]
Hi @timtebeek,I verified the fix with our Uber monorepo and discovered a couple more issues, which I’ve addressed with fixes in OpenRewrite (PR to be published soon). After running the FindMissingTypes recipe, I didn’t find any missing types.
Finally we got Lombok working! 🚀

**Comment by timtebeek** [comment]
Awesome, thanks a lot! I've merged & released your two most recent PRs, and look forward to the further fixes whenever ready. 🙏🏻

**Comment by amishra-u** [comment]
> Awesome, thanks a lot! I've merged & released your two most recent PRs, and look forward to the further fixes whenever ready. 🙏🏻

Thank you!

**Comment by timtebeek** [comment]
Also let us know when you think we can close this issue; I think the initial issue has been addressed, and we're waiting on:
- https://github.com/projectlombok/lombok/pull/3800

Anything left to do here then before we close this issue? Or can we close and monitor that upstream PR?

**Comment by amishra-u** [comment]
No, once #4835 and https://github.com/projectlombok/lombok/pull/3800 is merged. We will have lombok working for everything except few experimental ones.

There was some earlier activity on the PR from @Rawi01, but there hasn't been any recent activity—he might be on vacation. I'll post the PR to the Lombok forum for review next week.

**Comment by timtebeek** [comment]
Great to see your PR merged! A release might take a little while still, but at least there's good progress. I think from our side of things we could now look at adding the same Lombok handling to the Java 8 and 21 parsers as well; did you have that use case there too? Or should we pick that up on our end? Mostly to avoid any duplicate work

**Comment by amishra-u** [comment]
We don't have any usecase for  java-8 also not for java-21 yer. If there is no urgency I will take care this once I have some bandwidth.

**Comment by jevanlingen** [comment]
I will give it a go today for java-8 (I am at GMT+1).

**Comment by timtebeek** [comment]
Massive thanks again for the help here @amishra-u ! We ended up packaging an edge build of lombok to bring this to Java 11, 17 and 21 by default already.
- https://github.com/openrewrite/rewrite/pull/4989

Once there's an official release we can drop that local jar again; there's no further changes beyond pulling forward a newer build.

There's some work to be done still for Java 8 (fix missing type information & enable by default), but we can pick that up separately.
